_posts/2022-02-04-shard-indexing-backpressure-in-opensearch.markdown

```diff
@@ -0,0 +1,205 @@
+
+## **What is Shard Indexing Backpressure protection and Why is it important?**
+
+OpenSearch provides few control mechanisms today to prevent a data node meltdown for bulk indexing requests. These are essentially achieved by queue rejections, circuit breakers, and node-level memory limits for indexing traffic. However, these gating mechanisms are all static and reactive.
+
+* **Thread pool queue sizes** are static in nature and do not change dynamically in real time. Every incoming request in the queue consumes resources differently.
+* **Circuit Breakers** can track real-time memory utilization but most of them apply blanket level rejections without any fairness or identifying resources required for the requests.
+* **Node level Indexing Pressure limit** framework is designed to reject overwhelming indexing requests based on some pre-defined static limits on nodes. It is similar to the blanket level rejections offered by circuit breakers.
+
+Therefore, there is a need for smarter rejection mechanism at a granular level i.e. per shard request. This should help identify the underlying reason (slow indexing requests or stuck indexing tasks) for a performance breach in a cluster. Once the issue is detected, there is a need for an effective propagation of backpressure, transferring the load from the overwhelmed node to other nodes that are still healthy. Thus the healthy nodes are capable of taking smarter backoff and short circuit decisions while processing these requests.
+
+## How this feature solves the problem
+
+Shard Level Indexing Backpressure introduces dynamic rejections of indexing requests when there are too many stuck or slow requests in the cluster, breaching key performance thresholds. This prevents the nodes in cluster to run into cascading effects of failures under arising due to:
+
+1. **Performance Degradation –** Indexing performance for a shard starts to indicate slowness overtime, such as due to configuration changes, resource contention, or poison-pill requests.
+2. **Slow Nodes** – Entire data node starts to slowdown due to degraded hardware such as disk, network volumes, or software bugs such as thread-pool blocks, mutex locks.
+3. **Stuck Requests (Blackhole) -** Requests get stuck indefinitely and takes forever to responds back due to a slow downstream node. However the upstream node continues to accept requests faster than they are being processed.
+4. **Surge in traffic** – Sudden surge or spike in the traffic leading quick build up across the nodes.
```

Comment:
'traffic leading quick build up across the nodes.' -> 'traffic to a leading quick build up across the nodes.'